projects/gnutls/Dockerfile

```diff
@@ -0,0 +1,31 @@
+# Copyright 2016 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM ossfuzz/base-libfuzzer
+MAINTAINER alex.gaynor@gmail.com
+RUN apt-get install -y make autoconf automake libtool autopoint libnettle6 nettle-dev pkg-config gperf bison autogen texinfo curl
+
+RUN git clone https://gitlab.com/gnutls/gnutls.git
+RUN cd gnutls && git submodule update --init
+
+# Using the client_corpus transcripts from boringssl, they're a decent starting point.
```

Comment:
Ah, okay. Yeah, our PRNG is also deterministic in the fuzzers (see link). It means, if everything matches just right, you'll negotiate the same keys. But you can't modify anything around a crypto invariant. Encrypted records can't be modified. Signature checks likewise force you to leave everything signed unmodified. The handshake transcript bits mean your transcripts must match fuzzer's config exactly; any ClientHello deviance and transcript-based checks fail.

Most of the TLS 1.2 handshake is unencrypted though, so maybe deterministic PRNG is sufficient to explore most of it, renego and resumption aside, and assuming you always keep it up-to-date? TLS 1.3 is much more aggressively encrypted and transcript-sensitive, so I suspect that needs modification. Note: on our end, we still need to go analyze the coverage more carefully, so I'm mostly speculating.

(Note the deterministic PRNG also means everyone's corpus is specific to their implementation and PRNG. The other nice thing about fuzzer mode is transcripts become much more portable, assuming everyone agrees on what fuzzer mode means. Downside, of course, is you're not quite fuzzing the real code.)

By the way, since you're running against our test suite now, that has a `-transcript-dir` flag to get transcripts out. Take a look at the `fuzz/refresh_ssl_corpora.sh` script.